Fixing concurrency issue when generating C# function target.

fabiocav opened the pull request:
Fixing concurrency issue when generating C# function target and other minor enhancements to the invoker and generation process.

Resolves #92 

##### Issue description
The issue was caused by a race condition when `Invoke` was called concurrently on multiple threads after a function modification (where the cached function had been dropped), at that point, they ended up getting into the (previously named) `GetFunctionTarget` and the logic between `ReleaseContext` and `CreateContext` consumed enough time to let multiple threads attempt to set the assembly load context for a given function, a condition we were checking and preventing, which lead to the error.

##### Resolution
The issue actually uncovered the fact that we had a code path that was not thread safe being invoked by multiple threads. To resolve the problem, and address the issue without simply introducing locks that could live long enough to cause contention, the changes in this PR introduce a `FunctionValueLoader` class that lazily loads the function target when needed, caching and exposing the `Task` that actually performs the work. That `Task` will be returned to all subsequent callers, who will either efficiently await for the original task to complete or get the result of an already completed, cached `Task`. This new class builds on top of Lazy<T> and ensures single execution/publication, guaranteeing that only one thread will execute creation/compilation logic after the cache is dropped.
To deal with cache drops and creations that may happen while another creation is actively running, we are safely "replacing" the cache item and cancelling the pending operation (if one is running). Access to  the cache item are protected by a Reader/Writer lock in a short lived scope.

To increase reliability, the `FunctionAssemblyLoader` was also enhanced to use a concurrent dictionary and the API was modified to expose "atomic operations" when creating or updating the context.

mathewc commented:
Can you please explain what the issue was, how it occurred, and why these changes fix it?

mathewc commented on src/WebJobs.Script/Description/CSharp/CSharpFunctionInvoker.cs:
```diff
@@ -97,7 +101,7 @@ protected override void OnScriptFileChanged(object sender, FileSystemEventArgs e
+            ResetFunctionValue();
             TraceWriter.Verbose(string.Format(CultureInfo.InvariantCulture, "Script for function '{0}' changed. Reloading.", Metadata.Name));
 
             TraceWriter.Verbose("Compiling function script.");
```
I'm not sure the elapsed time we're outputting is something the user needs/wants to see. Early on it was useful for us to prove to ourselves that it was fast, but do we really need it in the logs each time?

fabiocav replied:
I agree. I'll remove that.

mathewc commented on src/WebJobs.Script/Description/CSharp/FunctionValueLoader.cs:
```diff
@@ -0,0 +1,51 @@
+
+namespace Microsoft.Azure.WebJobs.Script.Description
+{
+    internal sealed class FunctionValueLoader : Lazy<Task<MethodInfo>>, IDisposable
```
Please put a descriptive comment on this class. From reading the code, I'm not sure exactly why it is needed.

mathewc commented:
Also, since this change involves some tricky code, is there a simple repro/regression test that you can add so that we're confident we're not regressing when we make changes?

fabiocav commented:
I'll add some tests to cover the scenario this fixes as well as some of the more complex logic that was added/changed.

mathewc commented on src/WebJobs.Script/Description/CSharp/FunctionAssemblyLoader.cs:
```diff
@@ -15,11 +17,15 @@ namespace Microsoft.Azure.WebJobs.Script.Description
     public class FunctionAssemblyLoader : IDisposable
     {
-        private readonly List<FunctionAssemblyLoadContext> _functionContexts = new List<FunctionAssemblyLoadContext>();
+        public const string AssemblyPrefix = "\u0192-";
```
What is this prefix? Add some comments

fabiocav commented:
@mathewc I've updated the PR description with more details on the issue and information on the approach taken for the fix. Please let me know if you need additional details.

fabiocav commented:
@mathewc addressed your feedback and made some enhancements. I'm working on tests, but was wondering if we can have them in a separate PR and merge this in so we can have this changes ready for @davidebbo to deploy an update.

mathewc commented:
Sure, get it in :)
